query.go

```diff
@@ -778,120 +772,272 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
 	}
+	bestHeight := uint32(bestBlock.Height)
 
-	log.Debugf("Fetching filter for height=%v, hash=%v", height, blockHash)
+	// If the stop height with the maximum batch size is above our best
+	// known block, then we use the best block height instead.
+	stopHeight := startHeight + maxFilters - 1
+	if stopHeight > bestHeight {
+		stopHeight = bestHeight
+	}
 
-	// In addition to fetching the block header, we'll fetch the filter
-	// headers (for this particular filter type) from the database. These
-	// are required in order to verify the authenticity of the filter.
-	curHeader, err := getHeader(&blockHash)
+	stopHash, err := s.GetBlockHash(int64(stopHeight))
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't get cfheader for block %s "+
-			"from database", blockHash)
+		return nil, fmt.Errorf("unable to get hash for "+
+			"stopHeight=%d: %v", stopHeight, err)
 	}
-	prevHeader, err := getHeader(&block.PrevBlock)
+
+	// In order to verify the authenticity of the received filters, we'll
+	// fetch the block headers and filter headers in the range
+	// [startHeight-1, stopHeight]. We go one below our startHeight since
+	// the hash of the previous block is needed for validation.
+	numFilters := stopHeight - startHeight + 1
+	blockHeaders, _, err := s.BlockHeaders.FetchHeaderAncestors(
+		numFilters, stopHash,
+	)
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't get cfheader for block %s "+
-			"from database", blockHash)
+		return nil, fmt.Errorf("unable to get %d blockheader "+
+			"ancestors for stopHash=%d: %v", numFilters,
+			stopHash, err)
+	}
+
+	if len(blockHeaders) != int(numFilters)+1 {
+		return nil, fmt.Errorf("expected %d block headers, got %d",
+			numFilters+1, len(blockHeaders))
+	}
+
+	filterHeaders, _, err := s.RegFilterHeaders.FetchHeaderAncestors(
+		numFilters, stopHash,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get %d filter header "+
+			"ancestors for stopHash=%d: %v", numFilters, stopHash,
+			err)
+	}
+
+	if len(filterHeaders) != int(numFilters)+1 {
+		return nil, fmt.Errorf("expected %d filter headers, got %d",
+			numFilters+1, len(filterHeaders))
+	}
+
+	// We create a height index such that we can easily index into our
+	// header slices for a given block hash in the received response,
+	// without consulting the database. This also keeps track of for which
+	// blocks we are still awaiting a response. We start at index=1, as 0
+	// is for the block startHeight-1, only needed for validation.
+	heightIndex := make(map[chainhash.Hash]int)
+	for i := 1; i < len(blockHeaders); i++ {
+		block := blockHeaders[i]
+		heightIndex[block.BlockHash()] = i
+	}
+
+	// We'll immediately respond to the caller with the requested filter
+	// when it is received, so we make a channel to be notified when it's
+	// ready.
+	filterChan := make(chan *gcs.Filter, 1)
+
+	// We acquire the mutex ensuring we don't have several redundant
+	// CFilter queries running in parallel.
+	s.mtxCFilter.Lock()
```

Comment:
So if we have multiple goroutines trying to obtain the same set of filters, then at this point they've done a lot of unnecessary work w.r.t reading all the ancestors from disk, and obtaining the filter validation context.

Reply:
Fixed.